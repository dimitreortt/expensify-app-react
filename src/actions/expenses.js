import { db } from '../firebase/firebase'

export const expensesArrayToObject = (expenses) => {
  let expensesData = {}
  expenses.forEach(({ description, id, note, amount, createdAt }) => {
    expensesData[id] = { description, amount, note, createdAt }
  })
  return expensesData
}

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }

    return db.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

const expensesSnapshotToArray = (expensesSnapshot) => {
  let expenses = []
  expensesSnapshot.forEach((childExpenseSnapshot) => {
    expenses.push({
      id: childExpenseSnapshot.key,
      ...childExpenseSnapshot.val()
    })
  })
  return expenses
}

export const startSetExpenses = () => {
  // const expensesObject = expensesArrayToObject(expenses)
  return (dispatch) => {
    return db.ref('expenses').once('value').then((snapshot) => {
      const expenses = expensesSnapshotToArray(snapshot)
      dispatch(setExpenses(expenses))
    })
  }
}