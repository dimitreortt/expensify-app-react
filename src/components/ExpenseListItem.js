import React from 'react'
import { Link, NavLink } from 'react-router-dom'
import moment from 'moment'

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <NavLink to={'edit/' + id}>
      <h3>{description}</h3>
    </NavLink>
    <p>
      {amount}{' - '}
      {moment(createdAt).format('MMMM Do, YYYY')}
    </p>
  </div>
)

export default ExpenseListItem
