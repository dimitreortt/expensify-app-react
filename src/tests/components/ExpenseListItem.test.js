import expenses from "../fixtures/expenses"
import ExpenseListItem from "../../components/ExpenseListItem"
import { shallow } from "enzyme"
import React from 'react'

test('should test ExpenseListItem with given expense', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[1]} />)
  expect(wrapper).toMatchSnapshot()
})