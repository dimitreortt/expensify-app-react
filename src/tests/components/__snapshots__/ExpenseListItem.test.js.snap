// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test ExpenseListItem with given expense 1`] = `
<Link
  className="list-item"
  replace={false}
  to="edit/2"
>
  <div>
    <h3
      className="list-item__title"
    >
      Annuality
    </h3>
    <span
      className="list-item__sub-title"
    >
      December 27th, 1969
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $3,500.00
  </h3>
</Link>
`;
